### ISSUE TITLE
Strange NumberFormatException when having maps with Long keys

### ISSUE THREAD

**Issue by MrTetrixx** [flipkart-incubator/zjsonpatch#69]
Hello, I am using zjsonpatch in a project and I really like it.

I ran some of our data through it and came across a strange error. I have not looked at the zjsonpatch code yet but from the stacktrace it looks like it is doing `Integer.parseInt()` on that `Long` value used as a key in the map.

### The stacktrace:
```
Exception in thread "main" java.lang.NumberFormatException: For input string: "2407321151"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at com.flipkart.zjsonpatch.JsonDiff.isAllowed(JsonDiff.java:104)
	at com.flipkart.zjsonpatch.JsonDiff.introduceCopyOperation(JsonDiff.java:76)
	at com.flipkart.zjsonpatch.JsonDiff.asJson(JsonDiff.java:60)
	at com.flipkart.zjsonpatch.JsonDiff.asJson(JsonDiff.java:38)
	at JsonDiffTest.main(JsonDiffTest.java:33)
```

### Code to replicate the issue:
```java
public class JsonDiffTest {
    public static void main(String[] args) throws IOException {
        String a = "{\n" +
                "    \"map\": {\n" +
                "      \"3000000000\": {\n" +
                "        \"field\": 3100000000,\n" +
                "        \"otherField\": 0\n" +
                "      }\n" +
                "    }\n" +
                "  }";

        String b = "{\n" +
                "    \"map\": {\n" +
                "      \"3000000000\": {\n" +
                "        \"field\": 3100000000,\n" +
                "        \"otherField\": 0,\n" +
                "        \"extraField\": 0\n" +
                "      }\n" +
                "    }\n" +
                "  }";

        ObjectMapper mapper = new ObjectMapper();
        JsonNode jsonA = mapper.readTree(a);
        JsonNode jsonB = mapper.readTree(b);

        JsonDiff.asJson(jsonA, jsonB); // causes above stacktrace
    }
}
```

What makes it strange is that
* if you remove either of `otherField` or `extraField` it works.
* if you change the value of either of the fields with value `0` to anything else it works too.

**Comment by vishwakarma** [comment]
Thanks for raising the issue, please allow me a day to look into the issue, I will get back on it very very soon.

**Comment by vishwakarma** [comment]
@MrTetrixx This clearly a bug, line #104 & # 105, should be replaced by Long parsing ( this way it wont lose the necessary precision ).
Would you like to raise a PR for same or shall I take over on it?

**Comment by MrTetrixx** [comment]
You can take over, no problem.

Please check if there are any other issues of using unusual keys in a map, like BigDecimal or something.

I extended the test with a real assertion:
```
String a = "{" +
                "    \"map\": {" +
                "      \"3000000000\": {" +
                "        \"field\": 3100000000," +
                "        \"otherField\": 0" +
                "      }" +
                "    }" +
                "  }";

        String b = "{" +
                "    \"map\": {" +
                "      \"3000000000\": {" +
                "        \"field\": 3100000000," +
                "        \"otherField\": 0," +
                "        \"extraField\": 0" +
                "      }" +
                "    }" +
                "  }";

        JsonNode jsonA = objectMapper.readTree(a);
        JsonNode jsonB = objectMapper.readTree(b);

        JsonNode diff = JsonDiff.asJson(jsonA, jsonB);

        assertEquals(Operation.ADD.rfcName(), diff.get(0).get("op").textValue());
        assertEquals("/map/3000000000/extraField", diff.get(0).get("path").textValue());
        assertEquals("value", diff.get(0).get("value").textValue());
```
Should this test return a `copy` or an `add` operation? How does it decide between them?

**Comment by LouizFC** [comment]
@MrTetrixx copy comes over add when you create a field that has a value that is already present on your json.

In your case, you are creating a field called "extraField" with the same value as "otherField", so the copy takes "priority".

The library basically only generates Add and Remove operations if I am not wrong, then these operations are compacted into replace, move and copy operations. You can control this behavior using Flags

**Comment by MrTetrixx** [comment]
Yes, I guess I can set the `OMIT_COPY_OPERATION` and `OMIT_MOVE_OPERATION` flags to only get add, remove and replace.

**Comment by vishwakarma** [comment]
@MrTetrixx , i have put the fix in #70 , would you like to test it?

**Comment by MrTetrixx** [comment]
I have run our data through the diff again and it looks much better.
When will the fix be released?

**Comment by vishwakarma** [comment]
I am currently on vacation, I will release it on coming Thursday.

**Comment by MrTetrixx** [comment]
Great, looking forward to it!
(also stop checking github and enjoy your vacation :tropical_drink:)

**Comment by vishwakarma** [comment]
released 0.4.4, please use it.
Thank you for raising the bug
